src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -169,11 +166,39 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
     case o => wdlType.fromWdlString(dbValue)
   }
 
-  override def shutdown() = database.shutdown
+  /**
+    * Create a special execution context, a fixed thread pool, to run each of our composite database actions. Running
+    * each composite action as a runnable within the pool will ensure that-- at most-- the same number of actions are
+    * running as there are available connections. Thus there should never be a connection deadlock, as outlined in
+    * - https://github.com/slick/slick/issues/1274
+    * - https://groups.google.com/d/msg/scalaquery/5MCUnwaJ7U0/NLLMotX9BQAJ
+    *
+    * Custom future thread pool based on:
+    * - http://stackoverflow.com/questions/15285284/how-to-configure-a-fine-tuned-thread-pool-for-futures#comment23278672_15285441
+    *
+    * Database config parameter defaults based on: (expand the `forConfig` scaladoc for a full list of values)
+    * - http://slick.typesafe.com/doc/3.1.0/api/index.html#slick.jdbc.JdbcBackend$DatabaseFactoryDef@forConfig(path:String,config:com.typesafe.config.Config,driver:java.sql.Driver,classLoader:ClassLoader):JdbcBackend.this.Database
+    *
+    * Reuses the error reporter from the database's executionContext.
+    */
+  private val actionThreadPool: ExecutorService = {
+    val dbNumThreads = databaseConfig.getIntOr("db.numThreads", 20)
+    val dbMaximumPoolSize = databaseConfig.getIntOr("db.maxConnections", dbNumThreads * 5)
+    val actionThreadPoolSize = databaseConfig.getIntOr("actionThreadPoolSize", dbNumThreads) min dbMaximumPoolSize
+    Executors.newFixedThreadPool(actionThreadPoolSize)
+  }
+  private val actionExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(
+    actionThreadPool, database.executor.executionContext.reportFailure)
+
+  override def close(): Unit = {
+    actionThreadPool.shutdown()
+    database.close()
+  }
 
   // Run action with an outer transaction
   private def runTransaction[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect]): Future[R] = {
-    database.run(action.transactionally)
+    //database.run(action.transactionally) <-- https://github.com/slick/slick/issues/1274
+    Future(Await.result(database.run(action.transactionally), Duration.Inf))(actionExecutionContext)
```

Comment:
"custom" and "slick" in the branch name did not prepare me for this :sob:  Pardon me while I upvote slick/slick#1274